src/stdlib/lib.c

```diff
@@ -104,177 +137,170 @@ value_p printd(subrange_p whatever, subrange_p text) {
 }
 
 value_p extend_log10(subrange_p range) {
-	double val;
-	if(!assertSingle(range)) return new_val();
-	value_p initial = get_val(range, 0, 0);
-	val = log10(initial->numericVal);
-	value_p result = new_val();
-	setNumeric(result,val);
-	setFlag(result, FLAG_NUMBER);
-	return result;
+	if(!assertSingleNumber(range)) return new_val();
+	double initial = get_number(range);
+	double val = log10(initial);
+	return new_number(val);
 }
 
 value_p extend_sqrt(subrange_p range) {
-	double val;
-	if(!assertSingle(range)) return new_val();
-	value_p initial = get_val(range, 0, 0);
-	val = sqrt(initial->numericVal);
-	value_p result = new_val();
-	setNumeric(result,val);
-	setFlag(result, FLAG_NUMBER);
-	return result;
+	if(!assertSingleNumber(range)) return new_val();
+	double initial = get_number(range);
+	double val = sqrt(initial);
+	return new_number(val);
 }
 
 value_p extend_ceil(subrange_p range) {
-	double val;
-	if(!assertSingle(range)) return new_val();
-	value_p initial = get_val(range, 0, 0);
-	val = ceil(initial->numericVal);
-	value_p result = new_val();
-	setNumeric(result,val);
-	setFlag(result, FLAG_NUMBER);
-	return result;
+	if(!assertSingleNumber(range)) return new_val();
+	double initial = get_number(range);
+	double val = ceil(initial);
+	return new_number(val);
 }
 
 value_p extend_fabs(subrange_p range) {
-	double val;
-	if(!assertSingle(range)) return new_val();
-	value_p initial = get_val(range, 0, 0);
-	val = fabs(initial->numericVal);
-	value_p result = new_val();
-	setNumeric(result,val);
-	setFlag(result, FLAG_NUMBER);
-	return result;
+	if(!assertSingleNumber(range)) return new_val();
+	double initial = get_number(range);
+	double val = fabs(initial);
+	return new_number(val);
 }
 
 value_p extend_floor(subrange_p range) {
-	double val;
-	if(!assertSingle(range)) return new_val();
-	value_p initial = get_val(range, 0, 0);
-	val = floor(initial->numericVal);
-	value_p result = new_val();
-	setNumeric(result,val);
-	setFlag(result, FLAG_NUMBER);
+	if(!assertSingleNumber(range)) return new_val();
+	double initial = get_number(range);
+	double val = floor(initial);
+	return new_number(val);
+}
+
+value_p extend_open(subrange_p range_one, subrange_p range_two){
+	FILE *val;
+	if(!assertSingle(range_one) || !assertSingle(range_two) || open_num_files + 1 > MAX_FILES) return new_val();
+	value_p filename = get_val(range_one, 0, 0);
+	value_p mode = get_val(range_two, 0,0);
+	val = fopen(filename->str->text, mode->str->text);
+	if(val == NULL) return new_val();
+	open_num_files++;
+	open_files[open_num_files] = val;
+	return new_number((double) open_num_files);
+}
+
+value_p extend_close(subrange_p range){
+	if(!assertSingleNumber(range)) {
+		// Per the LRM this is actually supposed to crash the program.
+		fprintf(stderr, "EXITING - Attempted to close something that was not a valid file pointer\n");
+		exit(-1);
+	}
+
+	int fileNum = (int) get_number(range);
+	if (fileNum > open_num_files || open_files[fileNum] == NULL) {
```

Comment:
I don't see open_num_files ever being decremented, but if it was, this line could cause issues. Specifically use fileNum > MAX_FILES. Consider case file 1 and 2 are opened. 1 is closed, open_num_files is decremented to 1, then we get into the error condition.